language-server/src/references.js

```diff
@@ -0,0 +1,117 @@
+import { JsoncInstanceFromJSON } from "./json-instance-from-json.js";
+import { validate } from "./json-schema.js";
+import { JsoncInstance } from "./jsonc-instance.js";
+import { buildDiagnostic, extractBaseUri, isValidUrl } from "./util.js";
+
+// TODO: Handle anchor fragments
+
+/**
+ *
+ * @param {JsoncInstance} instance
+ * @returns
+ */
+export const validateReferences = async (instance) => {
+  const diagnostics = [];
+  const promises = [];
+  let baseUri = "";
+  /**
+   *
+   * @param {JsoncInstance} instance
+   * @param {string} basePath
+   * @returns
+   */
+  async function validateRefs(instance, basePath = "") {
+    if (instance.typeOf() === "object") {
+      for (const [key, valueInstance] of instance.entries()) {
+        if (key.value() == "$id" && typeof valueInstance.value() === "string") {
+          baseUri = extractBaseUri(valueInstance.value());
+        }
+        if (
+          key.value() === "$ref" &&
+          typeof valueInstance.value() === "string"
+        ) {
+          const ref = valueInstance.value();
+          const isLocalRef = isLocalReference(ref);
+          if (isLocalRef) {
+            const isValidRef = checkLocalReference(ref, instance);
+            if (!isValidRef) {
+              diagnostics.push(
+                buildDiagnostic(valueInstance, `Invalid reference: ${ref}`)
+              );
+            }
+            return;
+          }
+          promises.push(
+            handleExternalReference(ref, baseUri).then((isValidRef) => {
+              if (!isValidRef.success) {
+                diagnostics.push(
+                  buildDiagnostic(valueInstance, isValidRef.error)
+                );
+              }
+            })
+          );
+        } else {
+          validateRefs(valueInstance, basePath + "/" + key.value());
+        }
+      }
+    } else if (instance.typeOf() === "array") {
+      for (const item of instance.iter()) {
+        validateRefs(item, basePath);
+      }
+    }
+  }
+
+  validateRefs(instance);
+  await Promise.all(promises);
+
+  return diagnostics;
+};
+
+function isLocalReference(ref) {
+  return ref.startsWith("#");
+}
```

Comment:
This isn't going to be sufficient. What about a reference like, `another-schema#/nothing/here`?

Reply:
but `another-schema#/` is not a local reference right?
I couldn't find a example on the documentation which starts without `/` or `#` or a baseUri

Reply:
It's both. It's a reference to a location inside an external schema. You need to handle both it the schema doesn't exist and if the location doesn't exist within that external schema.